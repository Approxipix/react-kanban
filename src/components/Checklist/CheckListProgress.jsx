import React from 'react';
import styled from 'styled-components'

const Wrapper = styled.div`
  position: relative;
  margin-bottom: .75rem;
  padding-left: 2rem;
`;

const ProgressBar = styled.div`
  position: relative;
  height: .6rem;
  background-color: rgba(9,30,66,.08);
  border-radius: .5rem;
  overflow: hidden;
`;
const Progress = styled.div`
  position: absolute;
  display: block;
  width: 0;
  left: 0;
  top: 0;
  bottom: 0;
  background-color: ${props => !!props.color ? props.color : '#5ba4cf'};
  width: ${props => props.width}%;
  transition: width 0.2s ease-in-out;
`;

const Label = styled.div`
  position: absolute;
  top: -.25rem;
  right: calc(100% - 1.75rem);
  font-size: .75rem;
  color: #6b778c;
  white-space: nowrap;
`;

const CheckListProgress = (props) => {
  let color, width = 0;
  const
    { items } = props,
    allItems = items.length,
    doneItems = items.filter(item => !!item.status).length;

  if (!!allItems) width = (doneItems * 100) / allItems;
  if (allItems === doneItems) color = '#61bd4f';
  return (
    <Wrapper>
      <ProgressBar>
        <Progress width={width} color={color}/>
      </ProgressBar>
      <Label>{width.toFixed()} %</Label>
    </Wrapper>
  );
};

export default CheckListProgress;
