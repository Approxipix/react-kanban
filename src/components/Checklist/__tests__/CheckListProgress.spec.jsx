import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CheckListProgress from '../CheckListProgress';

describe('<CheckListProgress>', () => {
  let component;
  beforeEach(() => {
    const props = {
      items: [
        { status: true, description: 'Item 1' },
        { status: true, description: 'Item 2' },
        { status: false, description: 'Item 3' },
        { status: true, description: 'Item 4' },
        { status: false, description: 'Item 5' },
      ]
    };
    component = shallow(<CheckListProgress {...props} />);
  });

  it('should renders correctly without props', () => {
    const wrapper = shallow(<CheckListProgress />);

    expect(toJson(wrapper)).toMatchSnapshot()
  });

  it('should renders without crashing given the required props', () => {
    expect(toJson(component)).toMatchSnapshot()
  });

  it('checks if progress bar width sets correctly', () => {
    expect(component.find('Progress').props().width).toEqual(60);
  });

  it('should sets the color prop if all items done', () => {
    const props = {
      items: [
        { status: true, description: 'Item 1' },
        { status: true, description: 'Item 2' },
        { status: true, description: 'Item 3' },
      ]
    };
    const component = shallow(<CheckListProgress {...props} />);

    expect(component.find('Progress').props().color).toBeDefined();
  });
});
