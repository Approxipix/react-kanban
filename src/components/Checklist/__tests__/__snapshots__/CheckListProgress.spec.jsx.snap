// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckListProgress> should renders correctly without props 1`] = `""`;

exports[`<CheckListProgress> should renders without crashing given the required props 1`] = `
<styled.div>
  <styled.div>
    <Progress
      width={60}
    />
  </styled.div>
  <styled.div>
    60
     %
  </styled.div>
</styled.div>
`;
